src/loaddb/load_server_loader.cpp

```diff
@@ -114,8 +114,171 @@ namespace cubload
+	  int alloced_string = 0;
+
+	  for (i = 0; i < attr_info.num_values; i++)
+	    {
+	      error = or_get_attrname (&recdes, i, &string, &alloced_string);
+	      if (error != NO_ERROR)
+		{
+		  ASSERT_ERROR ();
+		  heap_attrinfo_end (&thread_ref, &attr_info);
+		  break;
+		}
+
+	      if (string != NULL && strcmp ("name", string) == 0)
+		{
+		  attr_idx = i;
+
+		  if (string != NULL && alloced_string == 1)
+		    {
+		      db_private_free_and_init (&thread_ref, string);
+		    }
+
+		  break;
+		}
+
+	      if (string != NULL && alloced_string == 1)
+		{
+		  db_private_free_and_init (&thread_ref, string);
+		}
+	    }
+	}
+
+	error = heap_scancache_end (&thread_ref, &scan_cache);
+	if (error != NO_ERROR)
+	  {
+	    ASSERT_ERROR ();
+	    heap_attrinfo_end (&thread_ref, &attr_info);
+	    return LC_CLASSNAME_ERROR;
+	  }
+
+	error = heap_get_class_info (&thread_ref, oid_User_class_oid, &hfid, NULL, NULL);
+	if (error != NO_ERROR)
+	  {
+	    ASSERT_ERROR ();
+	    heap_attrinfo_end (&thread_ref, &attr_info);
+	    return LC_CLASSNAME_ERROR;
+	  }
+
+	error = heap_scancache_start (&thread_ref, &scan_cache, &hfid, NULL, true, false, NULL);
+	if (error != NO_ERROR)
+	  {
+	    ASSERT_ERROR ();
+	    heap_attrinfo_end (&thread_ref, &attr_info);
+	    return LC_CLASSNAME_ERROR;
+	  }
+
+	while (true)
+	  {
+	    scan_code = heap_next (&thread_ref, &hfid, NULL, &inst_oid, &recdes, &scan_cache, PEEK);
+	    if (scan_code == S_SUCCESS)
+	      {
+		error = heap_attrinfo_read_dbvalues (&thread_ref, &inst_oid, &recdes, NULL, &attr_info);
+		if (error != NO_ERROR)
+		  {
+		    ASSERT_ERROR ();
+		    heap_scancache_end (&thread_ref, &scan_cache);
+		    heap_attrinfo_end (&thread_ref, &attr_info);
+		    return LC_CLASSNAME_ERROR;
+		  }
+
+		for (i = 0, heap_value = attr_info.values; i < attr_info.num_values; i++, heap_value++)
+		  {
+		    if (heap_value->attrid == attr_idx)
+		      {
+			const char *user_name = NULL;
+			char downcase_user_name[DB_MAX_USER_LENGTH] = { '\0' };
+			char user_specified_name[DB_MAX_IDENTIFIER_LENGTH] = { '\0' };
+
+			user_name = db_get_string (&heap_value->dbvalue);
+			intl_identifier_lower (user_name, downcase_user_name);
+			snprintf (user_specified_name, DB_MAX_IDENTIFIER_LENGTH, "%s.%s", downcase_user_name, class_name);
+
+			found_again = xlocator_find_class_oid (&thread_ref, user_specified_name, &class_oid, BU_LOCK);
+			if (found_again == LC_CLASSNAME_EXIST)
+			  {
+			    heap_scancache_end (&thread_ref, &scan_cache);
+			    heap_attrinfo_end (&thread_ref, &attr_info);
+			    return found_again;
+			  }
+
+			break;
+		      }
+		  }
+	      }
+	    else if (scan_code == S_END)
+	      {
+		break;
+	      }
+	    else
+	      {
+		ASSERT_ERROR ();
+		heap_scancache_end (&thread_ref, &scan_cache);
+		heap_attrinfo_end (&thread_ref, &attr_info);
+		return LC_CLASSNAME_ERROR;
+	      }
+	  }
+
+	heap_scancache_end (&thread_ref, &scan_cache);
+	heap_attrinfo_end (&thread_ref, &attr_info);
+      }
```

Comment:
please extract this code to a private function. If the public function is too long, It can be difficult to understand how this class works.

Reply:
Added the locate_class_for_all_users function to separate it.